File: strategy_core/context.py
"""Canonical strategy context contract shared across runtimes."""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable
from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from strategy_core.broker import Broker
    from strategy_core.capabilities import RuntimeCapabilities
    from strategy_core.data import StrategyDataClient
    from strategy_core.events import StrategyEvent
    from strategy_core.http import HttpClient
    from strategy_core.models import StrategyConfig
    from strategy_core.runtime import StrategyRuntime
    from strategy_core.state import MarketStateView
    from strategy_core.telemetry import Telemetry


@runtime_checkable
class StrategyContext(Protocol):
    """Runtime-neutral strategy authoring surface."""

    state: MarketStateView
    data: StrategyDataClient
    broker: Broker
    http: HttpClient
    runtime: StrategyRuntime
    capabilities: RuntimeCapabilities
    config: StrategyConfig
    telemetry: Telemetry

    def events(self) -> AsyncIterator[StrategyEvent]: ...


StrategyHandler = Callable[[StrategyContext], Awaitable[None]]

File: native/strategy_core/src/context.rs
use std::future::Future;

use crate::{
    Broker, HttpClient, MarketStateView, RuntimeCapabilities, StrategyConfig, StrategyDataClient,
    StrategyEvent, StrategyRuntime, Telemetry,
};

pub trait StrategyContext {
    type State: MarketStateView;
    type Data: StrategyDataClient;
    type Broker: Broker;
    type Http: HttpClient;
    type Runtime: StrategyRuntime;
    type Telemetry: Telemetry;

    fn state(&self) -> &Self::State;
    fn data(&self) -> &Self::Data;
    fn broker(&mut self) -> &mut Self::Broker;
    fn http(&self) -> &Self::Http;
    fn runtime(&mut self) -> &mut Self::Runtime;
    fn capabilities(&self) -> &RuntimeCapabilities;
    fn config(&self) -> &StrategyConfig;
    fn telemetry(&mut self) -> &mut Self::Telemetry;
    fn next_event(&mut self) -> impl Future<Output = Option<StrategyEvent>> + Send;
}

pub trait StrategyHandler<C: StrategyContext> {
    fn run(&mut self, ctx: &mut C) -> impl Future<Output = ()> + Send;
}
